crates/environ/src/vmoffsets.rs

```diff
@@ -358,7 +362,8 @@ impl<P: PtrSize> VMOffsets<P> {
 
         calculate_sizes! {
             typed_continuations_payloads: "typed continuations payloads object",
-            typed_continuations_store: "typed continuations store",
+            typed_continuations_stack_chain: "typed continuations stack chain",
+            typed_continuations_main_stack_limits: "typed continuations main stack limits",
```

Comment:
I don't understand your suggestion. What kind of shim object?
Currently, there is one `StackLimits` object per `ContinuationObject` (inlined in them) and one for the main stack. The one for the main stack lives at offset `typed_continuations_main_stack_limits` in the `VMContext`. If we inlined the main stack's `StackLimits` into the stack chain (by making the `MainStack` variant carry a `StackLimits` object directly, instead of a pointer to one), the type's size would grow from 2 words to 5. Since every `ContinuationObject` contains a `StackChain` object, this would waste a lot of space.
Moreover, the current design is deliberately designed such that we can uniformly access the `StackLimits` for `StackChain::Continuation` and `StackChain::MainStack` objects: Both contain a single pointer `p` as their payload. Since the `StackLimits` deliberately sits at offset 0 of the `ContinuationObject`, this means that `p` is always a pointer to the beginning of a `StacksLimit` object.

Reply:
What I mean is, why don't we have the invariant that `typed_continuation_stack_chain` is never `null`? I.e. we initialise it with a singleton chain containing the shim object for the main stack. To me it is a smell that we treat main stack differently from fiber stacks. We should ideally treat them uniformly, which would ultimately require us to alter the stack layout of the main stack slightly, though, in the mean time can at least make sure we don't treat it specially in the vmcontext.

Reply:
But that exact invariant is already established! If you look at the changes in `wasmtime_runtime::instance::Instance::initialize_vmctx` right after [here](https://github.com/wasmfx/wasmfxtime/pull/98/files#diff-e95cf87da8a8dcc08101777a3375ca3e634f0ed6b9f7c5daeec7e73cc4610eafL1135), you see that we initialize the `typed_continuations_stack_chain` field in the `VMContext` with `MainStack(p)`, where `p` points to the `VMContext`'s `typed_continuations_main_stack_limits` field.

Reply:
OK! Thanks for clarifying. I guess what I had envisaged was a design where we would have a `ContinuationObject`/`VMContRef` for the main stack too, such that stacks would be uniform. I suppose we can converge towards such a design later.